lsコマンドを作る4 の課題提出

88-99 opened the pull request:
lsコマンドを作る4 の課題を実装しました。
ご確認お願いいたします。

Commit:
lsコマンドを作る4 のfirst commit

Commit:
RuboCop指摘箇所の修正, メソッドの抽出

Commit:
permission実装部分の修正

summertree128 commented on 05.ls/lsk:
```diff
@@ -2,33 +2,74 @@
+def build_permission(stat)
+  permission_number = stat.mode.to_s(8)
+  permission_number.insert(0, '0') if permission_number.length == 5
+  file_type = { '01' => 'p', '02' => 'c', '04' => 'd', '06' => 'b', '10' => '-', '12' => 'l', '14' => 's' }
```
こういう感じで書いた方が見やすいかなと思いました。`file_mode`も同様です。

```suggestion
  file_type = {
    '01' => 'p',
    '02' => 'c',
    '04' => 'd',
    '06' => 'b',
    '10' => '-',
    '12' => 'l',
    '14' => 's'
  }
```

88-99 replied:
レビューありがとうございます。
修正いたしました。

summertree128 commented on 05.ls/lsk:
```diff
@@ -2,33 +2,74 @@
-  filenames_transposed = transpose_safely(filenames)
-  max_characters_of_filename = demand_max_characters_of_filename(filenames)
+  if argv_option[:l]
+    build_l_option(filenames)
```
`build_l_option`という名前のメソッドで`puts`までやってしまうのは少し責務が大きすぎるというか、名前と実態があっていないかなと思いました。

88-99 replied:
旧`build_l_option(filenames)`の処理の一部を別メソッドに分け、メソッド名を`build_and_show_row_l_option`に変更しました。`puts`は`show`で表現しました。

意図がくめているかわかりませんが最初考えたことは、`puts row`だけ`main`メソッドに移動させようと思いました。しかし`file_paths.each do |fp|`の中で`puts`しないと繰り返し処理を使っての表示ができませんでした。そのため`row`以外の処理を別メソッドに分けて旧`build_l_option`の処理内容を少なくし、メソッド名を処理に合わせる形にしました。
ご確認お願いいたします。

summertree128 replied:
@88-99 
修正ありがとうございます！
修正前よりはよくなったと思うのですが、ブロック数の方は`main`で`puts`しているのに残りの部分は`build_and_show_row_l_option`の中で出力までしている、ということでやはり粒度が揃っていない感じはあるかもですね。
コメントしていただいたように`puts`するところが`main`の中に来た方が良さそうです。
`build_and_show_row_l_option`を、`file_paths.each`の中で`puts`を繰り返すのではなく、`map`か何かで配列を作って返すようなメソッドにしてみてはどうでしょうか。配列の１要素に１行のデータが入っているイメージです。そうすれば`main`の方で`puts`できると思います。

88-99 replied:
修正いたしました。
配列を使えば良いのですね！
mapメソッドはeach文一回分の繰り返しが配列に入るイメージでしょうか。
勉強になりました😃ありがとうございます！

遠回りになりましたが、ブロック内が複雑に思えたので、以下を作ってからmapメソッドに変更しました。
```
rows = []
  file_paths.each do |fp|
    stat = File.stat(fp)
    row = "#{build_permission(stat)}  "
    row += "#{stat.nlink.to_s.rjust(file_nlinks.max.to_s.length)} "
    row += "#{Etc.getpwuid(File.stat(fp).uid).name}  "
    row += "#{Etc.getgrgid(File.stat(fp).gid).name}  "
    row += "#{stat.size.to_s.rjust(file_sizes.max.to_s.length)} "
    row += "#{stat.mtime.month.to_s.rjust(2)} "
    row += "#{stat.mtime.day.to_s.rjust(2)} "
    row += "#{stat.mtime.strftime('%H:%M')} "
    rows << row + File.basename(fp)     
  end
  rows
```

summertree128 commented on 05.ls/lsk:
```diff
@@ -2,33 +2,74 @@
+end
+
+def build_l_option(filenames)
+  file_path = filenames.map { |x| "#{Dir.getwd}/#{x}" }
+
+  file_nlink = file_path.map { |x| File.stat(x).nlink }
+
+  file_size = file_path.map { |x| File.stat(x).size }
```
この辺全て配列だと思うので、変数名は複数形の方がいいと思います。

88-99 replied:
忘れていました。修正いたしました。

summertree128 commented on 05.ls/lsk:
```diff
@@ -2,33 +2,74 @@
+    l_option_row += "#{stat.mtime.month.to_s.rjust(2)} "
+    l_option_row += "#{stat.mtime.day.to_s.rjust(2)} "
+    l_option_row += "#{stat.mtime.strftime('%H:%M')} "
+    l_option_row += "#{File.basename(fp)}\n"
```
あとで`puts`するので`\n`はいらないかなと思いましたがどうでしょうか？

88-99 replied:
`\n`なしで大丈夫でした。ありがとうございます。

summertree128 reviewed (commented):
レビューしました！簡潔でいい感じですね！
少しだけコメントしてあるのでご確認お願いしますー！

Commit:
メソッドの抽出とメソッド名の修正

hashを改行して表示すると縦に長くなるのでメソッドとして抽出しました。build_l_optionというメソッド名が処理の実態を表していないという指摘への対応として、処理の一部を別メソッドとして抽出し、メソッド名を処理内容に合わせるよう変更しました。

Commit:
putsをbuild_row_l_optionメソッドからmainメソッドに移動

mainメソッド内の処理の粒度を揃えるため

summertree128 reviewed (approved):
レビューOKです〜
